src/shadow-ui/prepare-stylesheet.js

```diff
@@ -0,0 +1,27 @@
+import css from 'css';
+import SHADOW_UI_CLASS_NAME from './class-name';
+
+const ID_AND_CLASS_REG_EXP = /#[a-zA-Z0-9_-]+|\.-?[a-zA-Z0-9_][a-zA-Z0-9_-]*/g;
+const ADD_POSTFIX_PATTERN  = '$&' + SHADOW_UI_CLASS_NAME.postfix;
+
+
+function addUIClassPostfix (rules) {
+    rules.forEach(rule => {
+        if (rule.type === 'rule' && rule.selectors && rule.selectors.length)
+            rule.selectors = rule.selectors.map(selector => selector.replace(ID_AND_CLASS_REG_EXP, ADD_POSTFIX_PATTERN));
+
+        if (rule.rules)
+            addUIClassPostfix(rule.rules);
```

Comment:
We need to avoid recursion here (in general, it's always a good rule). You can use stack technique here:
```js
// NOTE: make a shallow copy of array here, to avoid modification
// of the original `ast.stylesheet.rules`
var ruleStack = rules.slice();

while (ruleStack.length) {
   var rule = ruleStack.pop();
    // Do processing here
    ...

    if (rule.rules)
        ruleStack = ruleStack.concat(rule.rules);
}
```